### ISSUE TITLE
Print command not being called through gradient operation.

### ISSUE THREAD

**Issue by chasep255** [tensorflow/tensorflow#4224]
I have the following code.  When I call the image_graident function I expect it to print the image_reg variable.  It does not print anything when imageGraident() is called.

```
image_reg = tf.reduce_mean(tf.abs(input_image))
dream_obj = tf.nn.l2_loss(net.fc8) - image_reg
dream_obj = tf.Print(dream_obj, [image_reg])
delta_image = tf.gradients(dream_obj, input_image)[0]

delta_image = tf.Print(delta_image, [image_reg])

def imageGraident(img):
    orig_shape = img.shape
    img = np.reshape(img, (-1, 224, 224, 3))
    dimg = sess.run(delta_image, feed_dict = {input_image: img})
    return np.reshape(dimg, orig_shape)
```

However the following does work fine and it prints the value.

```
image_reg = tf.reduce_mean(tf.abs(input_image))
dream_obj = tf.nn.l2_loss(net.fc8) - image_reg

delta_image = tf.gradients(dream_obj, input_image)[0]
delta_image = tf.Print(delta_image, [image_reg])

def imageGraident(img):
    orig_shape = img.shape
    img = np.reshape(img, (-1, 224, 224, 3))
    dimg = sess.run(delta_image, feed_dict = {input_image: img})
    return np.reshape(dimg, orig_shape)
```

This seems like a bug to me but maybe I am missing something.

**Comment by tatatodd** [comment]
<wild_guess>
I'm guessing that this is because tf.Print is an identity op, which happens to have a side-effect of printing out the data in the second argument.  Identity ops don't have any impact on the gradient computation, and are either dropped when tf.gradients adds ops to the graph, or in some later optimization pass.
</wild_guess>

But @girving would know for sure.

**Comment by tatatodd** [comment]
@rmlarsen probably knows whether my explanation had any truth behind it.  Sorry for the redirects; I'm just getting ramped up on everything, and want to make sure the answer is correct.  Thanks!

**Comment by rmlarsen** [comment]
@tatatodd tf.Print does have a gradient op, which is inserted in the gradient subgraph, but it does not log anything:

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/logging_ops.py#L87

It is also true that ops that are not on a path from sink to source get pruned by the gradient computation:

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/gradients.py#L368

Does that answer your question?

**Comment by rmlarsen** [comment]
I think we can close this.

**Comment by joeyearsley** [comment]
Would it be possible to add this to the documentation of computer_gradient? - I've just spent excessive amount of time trying to debug why the print and control dependencies in my graph don't get executed.
